proof/refine/ARM/StateRelation.thy

```diff
@@ -880,33 +891,66 @@ lemma sc_replies_relation_prev_list:
   apply (clarsimp simp: sc_of_def opt_map_def projectKO_opt_sc split: option.splits)
   done
 
-(* FIXME: can generalise in terms of "sym heaps", maybe later *)
-lemma sym_refs_prev_eq_next_sc_replies:
-  "\<lbrakk> sym_refs (list_refs_of_replies' s');
-     heap_list (replyPrevs_of s') (scReply sc') replies\<rbrakk>
-    \<Longrightarrow> heap_list (replyNexts_of s') (scReply sc') (rev replies)"
-  oops
+lemma list_refs_of_replies'_reftype[simp]:
+  "(p, reftype) \<in> list_refs_of_replies' s' p' \<Longrightarrow> reftype \<in> {ReplyPrev, ReplyNext}"
+  by (clarsimp simp: list_refs_of_replies'_def list_refs_of_reply'_def get_refs_def2
+              split: option.split_asm)
+
+(* FIXME RT: move *)
+lemma opt_map_SomeD:
+  "f x = Some y \<Longrightarrow> (f |> g) x = g y"
+  by (clarsimp simp: opt_map_Some_def)
+
+lemma replyNext_replyNexts_of_opt_map[simp]:
+  "\<lbrakk>ksPSpace s' p = Some (KOReply reply); replyNext reply = Some (Next p')\<rbrakk>
+    \<Longrightarrow> (replyNexts_of s' |> replyPrevs_of s') p = replyPrevs_of s' p'"
+  by (clarsimp simp: opt_map_SomeD projectKO_opt_reply split: option.split)
+
+lemma sym_replies_next_then_prev_id:
+  "\<lbrakk>sym_refs (list_refs_of_replies' s');
+    ksPSpace s' p = Some (KOReply reply); replyNext_of reply = Some p'\<rbrakk>
```

Comment:
I would have expected a precondition of the form `replyNexts_of s' p = Some p'`. This would avoid unfolding this projection here and would fit in goals where you have the projection itself without forcing those to be unfolded as well (but with `in_opt_map_eq` in the simp set would also work for those where it is unfolded).

Reply:
Ah, you might still have to unfold it in this proof here for `sym_refs` (I'd suggest after `simp` with the right lemmas into `replyNexts_of s' |> replyPrevs_of s'`). Would still write it in contracted form in the lemma statement, though.

Reply:
I actually wrote both versions. You can find the projection version below. We can probably make use of both versions?

Reply:
Right, of course, should have realised that, sorry. I'd normalise to and keep only the contracted version, otherwise we'll start duplicating everything that has projections in it and will never we quite sure which version to use.